modules/tensor_mechanics/src/materials/ADComputeIncrementalShellStrain.C

```diff
@@ -0,0 +1,544 @@
+      1, Utility::string_to_enum<Order>(getParam<std::string>("order")));
+  _t_points = _t_qrule->get_points();
+  _strain_increment.resize(_t_points.size());
+  _total_strain.resize(_t_points.size());
+  _total_strain_old.resize(_t_points.size());
+  _B.resize(_t_points.size());
+  _B_old.resize(_t_points.size());
+  _ge.resize(_t_points.size());
+  _ge_old.resize(_t_points.size());
+  _Jmap.resize(_t_points.size());
+  _Jmap_old.resize(_t_points.size());
+  _dxyz_dxi.resize(_t_points.size());
+  _dxyz_deta.resize(_t_points.size());
+  _dxyz_dzeta.resize(_t_points.size());
+  _dxyz_dxi_old.resize(_t_points.size());
+  _dxyz_deta_old.resize(_t_points.size());
+  _dxyz_dzeta_old.resize(_t_points.size());
+  for (unsigned int i = 0; i < _t_points.size(); ++i)
+  {
+    _strain_increment[i] =
+        &declareADProperty<RankTwoTensor>("strain_increment_t_points_" + std::to_string(i));
+    _total_strain[i] =
+        &declareADProperty<RankTwoTensor>("total_strain_t_points_" + std::to_string(i));
+    _total_strain_old[i] =
+        &getMaterialPropertyOldByName<RankTwoTensor>("total_strain_t_points_" + std::to_string(i));
+    _B[i] = &declareADProperty<DenseMatrix<Real>>("B_matrix_t_points_" + std::to_string(i));
+    _B_old[i] =
+        &getMaterialPropertyOldByName<DenseMatrix<Real>>("B_matrix_t_points_" + std::to_string(i));
+    _ge[i] = &declareADProperty<RankTwoTensor>("ge_matrix_t_points_" + std::to_string(i));
+    _ge_old[i] =
+        &getMaterialPropertyOldByName<RankTwoTensor>("ge_matrix_t_points_" + std::to_string(i));
+    _Jmap[i] = &declareADProperty<Real>("J_mapping_t_points_" + std::to_string(i));
+    _Jmap_old[i] = &getMaterialPropertyOldByName<Real>("J_mapping_t_points_" + std::to_string(i));
+    _dxyz_dxi[i] = &declareADProperty<RealVectorValue>("dxyz_dxi_t_points_" + std::to_string(i));
+    _dxyz_dxi_old[i] =
+        &getMaterialPropertyOldByName<RealVectorValue>("dxyz_dxi_t_points_" + std::to_string(i));
+    _dxyz_deta[i] = &declareADProperty<RealVectorValue>("dxyz_deta_t_points_" + std::to_string(i));
+    _dxyz_deta_old[i] =
+        &getMaterialPropertyOldByName<RealVectorValue>("dxyz_deta_t_points_" + std::to_string(i));
+    _dxyz_dzeta[i] =
+        &declareADProperty<RealVectorValue>("dxyz_dzeta_t_points_" + std::to_string(i));
+    _dxyz_dzeta_old[i] =
+        &getMaterialPropertyOldByName<RealVectorValue>("dxyz_dzeta_t_points_" + std::to_string(i));
+  }
+}
+
+template <ComputeStage compute_stage>
+void
+ADComputeIncrementalShellStrain<compute_stage>::initQpStatefulProperties()
+{
+  computeGMatrix();
+  computeBMatrix();
+}
+
+template <ComputeStage compute_stage>
+void
+ADComputeIncrementalShellStrain<compute_stage>::computeProperties()
+{
+  // quadrature points in isoparametric space
+  _2d_points = _qrule->get_points(); // would be in 2D
+
+  unsigned int dim = _current_elem->dim();
+  if ((dim != 2))
+    mooseError("Shell element is implemented only for 2D Linear elements");
```

Comment:
Same issues I mentioned before about this check -- do it once upfront if possible, and check for more things